compiler/shared/test/diff/Defunctionalize/OldMonoList.mls

```diff
@@ -0,0 +1,275 @@
+//│ class Lambda1$2$3()
+//│ fun map$List$1: (anything, anything) -> error
+//│ fun add2$1: Int -> Int
+//│ fun main$$5: () -> error
+//│ fun apply$Lambda1$3$4: (anything, Int) -> Int
+//│ fun map$Nil$2: forall 'a. ('a, anything) -> 'a
+//│ fun apply$Lambda1$2$3: (anything, Int) -> Int
+//│ error
+//│ Code generation encountered an error:
+//│   if expression was not desugared
+
+:e //FIXME: Class parameters not lifted
+:mono
+:AllowRuntimeErrors
+class List(e: Int, tail: List | Nil) {
+  fun count(): Int
+  fun count() = 1 + tail.count()
+}
+class Nil() {
+  fun count() = 0
+}
+fun foo(x) = x.count()
+fun generate(x) = 
+  if x > 0 then new List(x, generate(x+1)) else new Nil()
+foo(List(1, List(2, Nil())))
+foo(generate(1))
+//│ Lifted:
+//│ TypingUnit {
+//│   class List$1([e: Int, tail: |(List, Nil,),]) {
+//│     fun count = () -> Int
+//│     fun count = () => +(1, ((this).tail).count(),)
+//│   }
+//│   class Nil$2([]) {fun count = () => 0}
+//│   fun foo$1 = (x,) => (x).count()
+//│   fun generate$2 = (x,) => {if (>(x, 0,)) then new List$1([x, generate$2(+(x, 1,),),]) {} else new Nil$2([]) {}}
+//│   Code(List(foo$1(List$1(1, List$1(2, Nil$2(),),),)))
+//│   Code(List(foo$1(generate$2(1,),)))
+//│ }
+//│ Mono:
+//│ TypingUnit {
+//│   class Nil$2([]) {}
+//│   class List$1([e: Int, tail: |(List, Nil,),]) {}
+//│   fun foo$1 = (x,) => let obj = x in if obj is ‹(Nil$2) then count$Nil$2(obj,); (List$1) then count$List$1(obj,); else error›
+//│   fun count$Nil$2 = (this,) => 0
+//│   fun count$List$1 = (this,) => +(1, let obj = let obj = this in if obj is ‹(List$1(e, tail,)) then tail; else error› in if obj is ‹(Nil$2) then count$Nil$2(obj,); (List$1) then count$List$1(obj,); else error›,)
+//│   fun generate$2 = (x,) => {if (>(x, 0,)) then new List$1([x, generate$2(+(x, 1,),),]) {} else new Nil$2([]) {}}
+//│   fun main$$5 = () => foo$1(generate$2(1,),)
+//│   fun main$$4 = () => foo$1(List$1(1, List$1(2, Nil$2(),),),)
+//│   Code(List(main$$4()))
+//│   Code(List(main$$5()))
+//│ }
+//│ ╔══[ERROR] type identifier not found: List
+//│ ║  l.74: 	class List(e: Int, tail: List | Nil) {
+//│ ╙──      	                         ^^^^
+//│ ╔══[ERROR] type identifier not found: Nil
+//│ ║  l.74: 	class List(e: Int, tail: List | Nil) {
+//│ ╙──      	                                ^^^
+//│ ╔══[ERROR] Currently unsupported `new` syntax
+//│ ║  l.83: 	  if x > 0 then new List(x, generate(x+1)) else new Nil()
+//│ ╙──      	                         ^^^^^^^^^^^^^^^
+//│ ╔══[ERROR] Currently unsupported `new` syntax
+//│ ╙──
+//│ class Nil$2()
+//│ class List$1(e: Int, tail: error)
+//│ fun foo$1: Object -> Int
+//│ fun count$Nil$2: anything -> 0
+//│ fun count$List$1: Object -> Int
+//│ fun generate$2: Num -> error
+//│ fun main$$5: () -> Int
+//│ fun main$$4: () -> Int
+//│ Int
+//│ res
+//│     = 2
+//│ res
+//│ Runtime error:
+//│   RangeError: Maximum call stack size exceeded
+
+:e //FIXME: Class parameters not lifted
```

Comment:
Shouldn't this be fixed before we merge the PR?

Reply:
Fixed.